08-docker/README.md

```diff
@@ -0,0 +1,252 @@
+- **Docker Registry** is a centralized location for storing and distributing Docker images. The most commonly used public registry is **Docker Hub**, but you can also create your own private registry.
+- **Docker Hub** is a Docker Registry that provides unlimited storage for public images and offers paid plans for hosting private images. Anybody can access a public image. But to publish and access images on Docker Hub, you must create an account first.
+
+## Docker vs Virtual Machines
+
+![1](../resources/vm_vs_containers.png)
+
+## Create Docker network for spring petclinic containers <a name="network"></a>
+
+By default, Docker uses a bridge network with a subnet like 172.17.0.0/16 and a gateway like 172.17.0.1 for containers to communicate with the host and other containers on the same network. We can create our custom network,for example:
+
+```
+docker network create petclinic-network
+```
+
+Docker automatically assign a subnet (`172.18.0.0/16`) and gateway (`172.18.0.1`) to the network if not specified. 
+
+Use `--network petclinic_network` with `run` command when creating a container or connect a running container to a network with `docker network connect multi-host-network container1`. For example: `docker network connect petclinic_network petclinic-rest`
+
+## Add volume to PostgreSQL docker container <a name="volume"></a>
+
+The purpose of using Docker volumes is to persist data outside the container so it can be backed up or shared. Docker volumes are dependent on Docker's file system and are the preferred method of persisting data for Docker containers and services.
+
+There are three types of volumes: anonymous, named, and host. Docker manages both anonymous and named volumes, automatically mounting them in self-generated directories in the host. While anonymous volumes were useful with older versions of Docker (pre 1.9), **named ones are now the suggested way to go**. Host volumes also allow us to specify an existing folder in the host.
+
+We can configure host volumes at the service level, and named volumes in the outer level of the configuration, in order to make the latter visible to other containers, rather than only to the one they belong. Multiple containers can use the same volume in the same time period. This is useful if two containers need access to shared data. For example, if one container writes and the other reads the data. Volume names must be unique among drivers. This means you cannot use the same volume name with two different drivers.
+
+It is not required to first create volume: `docker volume create [OPTIONS] [VOLUME]`. You can also specify e.g. `-v psql-db:/var/lib/postgresql/data` it in `docker run` command:
+```
+docker run -d -p 5432:5432 --network petclinic_network -v psql-db:/var/lib/postgresql/data --name petclinic-postgres petclinic-postgres
```

Comment:
In previous step network name is `petclinic-network` with dash (-). Here, underscore (_) is provided.